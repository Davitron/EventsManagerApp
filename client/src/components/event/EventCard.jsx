import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { Card, Image, Button, Label } from 'semantic-ui-react';
import PropTypes from 'prop-types';

const getColor = (status) => {
  if (status === 'pending') return 'orange';
  if (status === 'cancelled') return 'red';
  if (status === 'accepted') return 'teal';
};

const handleDuration = (sDate, eDate) => {
  sDate = moment(sDate).format('MMMM Do YYYY');
  eDate = moment(eDate).format('MMMM Do YYYY');
  if (sDate === eDate) {
    return sDate;
  }
  return `${sDate} - ${eDate}`;
};

const EventCard = ({
  event,
  mode,
  onPositive,
  onNegative
}) => (
  <Card>
    <Image src={event.image} />
    <Card.Content>
      <Label color={getColor(event.status)} ribbon="right">{event.status}</Label>
      <Card.Header>{event.eventName}</Card.Header>
      <Card.Meta>{handleDuration(event.startDate, event.endDate)}</Card.Meta>
      <Card.Description><Link to={`/centers/${event.centerId}`}>{event.center.name}</Link></Card.Description>
    </Card.Content>
    <Card.Content extra>
      {
        (onNegative && onPositive) &&
        <div className="ui two buttons">
          <Button basic onClick={onPositive} color="green">{ mode ? 'Accept' : 'Update' }</Button>
          <Button basic onClick={onNegative} color="red">{ mode ? 'Decline' : 'Delete' }</Button>
        </div>
      }
    </Card.Content>
  </Card>
);

EventCard.propTypes = {
  event: PropTypes.objectOf(() => null).isRequired,
  mode: PropTypes.string,
  onPositive: PropTypes.func,
  onNegative: PropTypes.func
};

EventCard.defaultProps = {
  onPositive: null,
  onNegative: null,
  mode: null
};

export default EventCard;
