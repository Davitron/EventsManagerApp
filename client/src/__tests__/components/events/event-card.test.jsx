import React from 'react';
import expect from 'expect';
import { shallow } from 'enzyme';
import { Card } from 'semantic-ui-react';
import EventCard from '../../../components/event/event-card';

let mountedComponent;
const dispatch = fn => fn;

let props;
const locations = [];
const history = {
  push(location) {
    locations.push(location);
  },
};

const eventObj = {
  eventName: 'Test Arena',
  centerId: 2,
  days: '2',
  image: {
    files: []
  },
  status: 'pending',
  Center: {
    name: 'test center'
  },
  startDate: '2018-12-12'
};

const updateEvent = jest.fn();
const deleteEvent = id => id;


const location = {};

let wrapper;

/**
 * @description Initialise the component
 *
 * @param {object} event
 * @param {string} mode
 * @returns {object} mountedComponent - Mounte
 */
const getComponent = (event, mode) => {
  if (!mountedComponent) {
    props = {
      dispatch,
      history,
      location,
      onPositive: updateEvent,
      onNegative: deleteEvent,
      mode,
      event
    };
    history.push = jest.fn();
    mountedComponent = shallow(<EventCard {...props} />);
  }
  return mountedComponent;
};
describe('Event Component', () => {
  beforeAll(() => {
  });


  it('component successfully rendered', () => {
    wrapper = getComponent(eventObj, 'Accept');
    expect(wrapper).toMatchSnapshot();
  });

  it('testing get color function', () => {
    const newEvent = {
      ...eventObj,
      status: 'cancelled'
    };
    const card = getComponent(newEvent).find(Card);
    expect(card.length).toBe(1);
  });
});

