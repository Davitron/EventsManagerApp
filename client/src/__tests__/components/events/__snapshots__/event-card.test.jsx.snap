// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Component component successfully rendered 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EventCard
    dispatch={[Function]}
    event={
      Object {
        "Center": Object {
          "name": "test center",
        },
        "centerId": 2,
        "days": "2",
        "eventName": "Test Arena",
        "image": Object {
          "files": Array [],
        },
        "startDate": "2018-12-12",
        "status": "pending",
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={Object {}}
    mode="Accept"
    onNegative={[Function]}
    onPositive={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Image
          as="img"
          src={
            Object {
              "files": Array [],
            }
          }
          ui={true}
        />,
        <CardContent>
          <Label
            color="orange"
            ribbon="right"
          >
            pending
          </Label>
          <CardHeader>
            Test Arena
          </CardHeader>
          <CardMeta>
            December 12th 2018 - June 11th 2018
          </CardMeta>
          <CardDescription>
            <Link
              replace={false}
              to="/centers/2"
            />
          </CardDescription>
        </CardContent>,
        <CardContent
          extra={true}
        >
          <div
            className="ui two buttons"
          >
            <Button
              as="button"
              basic={true}
              color="green"
              onClick={[MockFunction]}
            >
              Accept
            </Button>
            <Button
              as="button"
              basic={true}
              color="red"
              onClick={[Function]}
            >
              Decline
            </Button>
          </div>
        </CardContent>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "img",
          "src": Object {
            "files": Array [],
          },
          "ui": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Label
              color="orange"
              ribbon="right"
            >
              pending
            </Label>,
            <CardHeader>
              Test Arena
            </CardHeader>,
            <CardMeta>
              December 12th 2018 - June 11th 2018
            </CardMeta>,
            <CardDescription>
              <Link
                replace={false}
                to="/centers/2"
              />
            </CardDescription>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "pending",
              "color": "orange",
              "ribbon": "right",
            },
            "ref": null,
            "rendered": "pending",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Test Arena",
            },
            "ref": null,
            "rendered": "Test Arena",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "December 12th 2018 - June 11th 2018",
            },
            "ref": null,
            "rendered": "December 12th 2018 - June 11th 2018",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Link
                replace={false}
                to="/centers/2"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": undefined,
                "replace": false,
                "to": "/centers/2",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div
            className="ui two buttons"
          >
            <Button
              as="button"
              basic={true}
              color="green"
              onClick={[MockFunction]}
            >
              Accept
            </Button>
            <Button
              as="button"
              basic={true}
              color="red"
              onClick={[Function]}
            >
              Decline
            </Button>
          </div>,
          "extra": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                as="button"
                basic={true}
                color="green"
                onClick={[MockFunction]}
              >
                Accept
              </Button>,
              <Button
                as="button"
                basic={true}
                color="red"
                onClick={[Function]}
              >
                Decline
              </Button>,
            ],
            "className": "ui two buttons",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "basic": true,
                "children": "Accept",
                "color": "green",
                "onClick": [MockFunction],
              },
              "ref": null,
              "rendered": "Accept",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "basic": true,
                "children": "Decline",
                "color": "red",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Decline",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Image
            as="img"
            src={
              Object {
                "files": Array [],
              }
            }
            ui={true}
          />,
          <CardContent>
            <Label
              color="orange"
              ribbon="right"
            >
              pending
            </Label>
            <CardHeader>
              Test Arena
            </CardHeader>
            <CardMeta>
              December 12th 2018 - June 11th 2018
            </CardMeta>
            <CardDescription>
              <Link
                replace={false}
                to="/centers/2"
              />
            </CardDescription>
          </CardContent>,
          <CardContent
            extra={true}
          >
            <div
              className="ui two buttons"
            >
              <Button
                as="button"
                basic={true}
                color="green"
                onClick={[MockFunction]}
              >
                Accept
              </Button>
              <Button
                as="button"
                basic={true}
                color="red"
                onClick={[Function]}
              >
                Decline
              </Button>
            </div>
          </CardContent>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "img",
            "src": Object {
              "files": Array [],
            },
            "ui": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Label
                color="orange"
                ribbon="right"
              >
                pending
              </Label>,
              <CardHeader>
                Test Arena
              </CardHeader>,
              <CardMeta>
                December 12th 2018 - June 11th 2018
              </CardMeta>,
              <CardDescription>
                <Link
                  replace={false}
                  to="/centers/2"
                />
              </CardDescription>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "pending",
                "color": "orange",
                "ribbon": "right",
              },
              "ref": null,
              "rendered": "pending",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Test Arena",
              },
              "ref": null,
              "rendered": "Test Arena",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "December 12th 2018 - June 11th 2018",
              },
              "ref": null,
              "rendered": "December 12th 2018 - June 11th 2018",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Link
                  replace={false}
                  to="/centers/2"
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": undefined,
                  "replace": false,
                  "to": "/centers/2",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div
              className="ui two buttons"
            >
              <Button
                as="button"
                basic={true}
                color="green"
                onClick={[MockFunction]}
              >
                Accept
              </Button>
              <Button
                as="button"
                basic={true}
                color="red"
                onClick={[Function]}
              >
                Decline
              </Button>
            </div>,
            "extra": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Button
                  as="button"
                  basic={true}
                  color="green"
                  onClick={[MockFunction]}
                >
                  Accept
                </Button>,
                <Button
                  as="button"
                  basic={true}
                  color="red"
                  onClick={[Function]}
                >
                  Decline
                </Button>,
              ],
              "className": "ui two buttons",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "basic": true,
                  "children": "Accept",
                  "color": "green",
                  "onClick": [MockFunction],
                },
                "ref": null,
                "rendered": "Accept",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "basic": true,
                  "children": "Decline",
                  "color": "red",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Decline",
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
